### ISSUE TITLE
proposal to set up language-agnostic testing of inheritance modes

### ISSUE THREAD

**Issue by brentp** [arq5x/gemini#734]
Putting this here in gemini issues for lack of a better place ...

Testing variants under different inheritance modes is hard.

It's implemented in multiple places including

+ gemini
+ xbrowse
+ [genmod](http://moonso.github.io/genmod/)

and likely more.

I propose to make a **simple-to-parse, programming-language and implementation-agnostic text format to be used for testing inheritance tools**.

Minimally, the format must indicate:
+ genotypes
+ family structure
+ inheritance mode(s) that should be determined by the tool

It will support only single families.

As a first draft in json format would look like this:

```
{
'name_of_test':
	{'genotypes': ['0/1', '0/1', '1/1'],
	 'pedigree': ['dad:u:m', 'mom:u:f', 'kid:a:f:dad:mom'],
	 'modes': ['autosomal_recessive'],
	 'comment': "simple autosomal-recessive test"
	},
'dominant-1':
	 {'genotypes': ['0/0', '0/1', '0/1'],
	 'pedigree': ['dad:u:m', 'mom:a:f', 'kid:a:f:dad:mom'],
	 'modes': ['autosomal_dominant', 'x-linked-dominant'],
	 'comment': "autosomal and x-linked-dominant"
	},
}
```

where the format in *pedigree* is a flat list where each sample is
essentially a condensed line from a ped file with ':'-delimited values
of:

```
+ sample_id
+ affection-status: 'u'naffected, 'a'ffected, 'c'arrier '?'unknown
+ sex: 'm'ale, 'f'emale, 'u'nknown
+ dad-id (optional)
+ mom-id (optional)
```

This will be more concise than actual ped format, but still allow specifying
multiple generation pedigrees.

Modes indicate what patterns the proposed test meets. Available modes are:

+ 'autosomal-dominant'
+ 'autosomal-recessive'
+ 'de-novo'
+ 'x-linked-recessive'
+ 'x-linked-dominant'
+ 'x-linked-denovo'
+ 'compount-het' # will require a nested list of genotypes.

With this in place, any implementation can use this (hopefully) community-driven and -vetted
set of tests.

cc: @bw2 @moonso @jxchong @arq5x

If anyone is interested, perhaps we can start a new repo and have peer-review on pull-requests
that add test-cases to the above JSON.

The repo will consist only of the json tests and docs in the README.

We are planning to do this anyway for gemini as we support expanding genotypes into separate
tables so we will need a set of implementation-independent tests.

Please let me know any comments and CC anyone else who might be interested.

-Brent

**Comment by moonso** [comment]
Excellent initiative! I would like to contribute to this.
Would be nice to have discussions about how to handle some of the more complicated models, missing genotypes etc.

Could you be a bit more explicit how a test run could look like? 
Why not use simulated ped- and vcf-files?

Måns

**Comment by brentp** [comment]
great!

I test run is handled internally by the implementation. This would just be a collection of the complicated (and simple) edge-cases.

The justification for **not** having a VCF and PED for everything is that it's onerous to create a PED and VCF for every case but simpler and easier to see all the information in 1 place with an abbreviated format. When we get bug reports on gemini, it often takes me a long time to understand everything that's going on in the VCF and ped, but if it can be distilled down to genotypes and a simple family structure it's more easily resolved.

That said, I do see the benefit to using the canonical means of representing these in PED and VCF and could be convinced to use them instead.

**Comment by moonso** [comment]
Sure I agree on the problem of searching for errors in the vcf.
So if we have a well defined json format we could add a small script in this new package that converts the json to .ped and .vcf?

**Comment by brentp** [comment]
ah, yes, that's a great idea.

**Comment by brentp** [comment]
I have been revisiting this again with Tom Nicholas in the lab. We were considering if there's a way to represent the genotypes after decomposition that would simplify a lot of the problems we address in gemini.

Let's Say we have a tri-allelic variant: REF A, ALT: T, G, C

| REF: A | Mom (A/T) | Dad (T/G) | Kid (A/C) |
| ---------  | -------------- | --------------| -------------| 
|     T       |     0/1        |      ./1        |    0/.         |
|     G       |    0/.          |      ./1        |    0/.        |
|     C      |    0/.           |       ./.        |   0/1         |

Here, the kid has the novel allele `C` which should be called de novo. If we represent the dad's A/C as Unknown, this will certainly be missed. However, in this case, we (can) know that the dad's variant came from a decomposed allele and therefore it's "valid". If we represent the data as the number of alternate alleles, then for A/C, the dad has 0, the mom has 0 and the kid has 1, so this must be de novo.

The difficulty is that we must differentiate unknown by decomposition (which we can convert to 0 alternates) from unknown because of lack of sequence depth. We can know this from the FORMAT DP field which both freebayes and GATK report with Number=1 so it indicates the total depth at that locus. If that value is 0 or unknown, then we must leave a ./. genotype as unknown. If it is > 0, then we can assume the ./. genotype came from decomposition and we can set ./. to HOM_REF (or to 0 alternate alleles). This adds the difficulty of encoding that we must also check DP, but we can make this a one-time cost when creating a new database. **Please let me know if there are any issues I am not seeing with this approach.**

To summarize, we encode:
+ 0/0, ./0, ./. (with depth 0), 0/. as 0
+ 0/1, ./1 as 1 (currently this is represented as unknown)
+ 1/1 as 2

Given this encoding, we could instead write the 'genotypes' field in the proposed JSON for Kid, Mom, Dad as:

'genotypes': [1, 0, 0],

From this, it's easy to see by eye that the Kid has an alternate that is not present in mom and dad. In contrast, a missing value would be:

'genotypes': [-1, 0, 0],

We could also easily find other odd denovos like:

'genotypes': [2, 1, 0]

where, when there are no unknowns, the rule is simply: `denovo == (Kid > Mom + Dad) || (2 - Kid) > (2 - Mom) + (2 - Dad)` 

The (above proposed) program that writes a VCF from the JSON could output a variant with ./1 and 0/1 for the `1` encoding.

**Comment by brentp** [comment]
I have a start for this here: https://github.com/quinlan-lab/mendacity
